cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -416,22 +448,64 @@ String podName(Integer podId) {
      * @return A future which completes the the node id of the controller of the cluster,
      * or -1 if there is not currently a controller.
      */
-    int controller(int podId, Admin ac, long timeout, TimeUnit unit) throws ForceableProblem, InterruptedException {
+    int controller(int podId, Admin ac, long timeout, TimeUnit unit, RestartContext restartContext) throws ForceableProblem, InterruptedException {
         Node controllerNode = null;
         try {
             DescribeClusterResult describeClusterResult = ac.describeCluster();
             KafkaFuture<Node> controller = describeClusterResult.controller();
             controllerNode = controller.get(timeout, unit);
+            restartContext.clearConnectionError();
         } catch (ExecutionException e) {
-            throw new ForceableProblem("Error while trying to determine the cluster controller from pod " + podName(podId), e.getCause());
+            maybeTcpProbe(podId, e, restartContext);
         } catch (TimeoutException e) {
-            throw new ForceableProblem("Error while trying to determine the cluster controller from pod " + podName(podId), e);
+            throw new ForceableProblem("lala Error while trying to determine the cluster controller from pod " + podName(podId), e);
         }
         int id = controllerNode == null || Node.noNode().equals(controllerNode) ? -1 : controllerNode.id();
         log.debug("controller is {}", id);
         return id;
     }
 
+    /**
+     * If we've already had trouble connecting to this broker try to probe whether the connection is
+     * open on the broker; if it's not then maybe throw a ForceableProblem to immediately force a restart.
+     * This is an optimization for brokers which don't seem to be running.
+     */
+    private void maybeTcpProbe(int podId, ExecutionException executionException, RestartContext restartContext) throws ForceableProblem {
+        log.info("debuug1 {} {}", restartContext.connectionError(), System.currentTimeMillis());
+        if (restartContext.connectionError() + 120_000 >= System.currentTimeMillis()) {
```

Comment:
Why 120_000 an not operationTimeoutMs?

Reply:
It is related to AdminClient default timeout (1 minute) + some reasonable offset.